pylint/testutils/functional/find_functional_tests.py

```diff
@@ -62,17 +62,21 @@ def _check_functional_tests_structure(directory: Path) -> None:
     dirs: set[Path] = set()
 
     def walk(path: Path) -> Iterator[Path]:
+        violations: list[tuple[Path, int]] = []
         for _file_or_dir in path.iterdir():
             if _file_or_dir.is_dir():
                 _files = list(_file_or_dir.iterdir())
-                assert len(_files) <= REASONABLY_DISPLAYABLE_VERTICALLY, (
-                    f"{_file_or_dir} contains too many functional tests files "
-                    + f"({len(_files)} > {REASONABLY_DISPLAYABLE_VERTICALLY})."
-                )
+                if len(_files) > REASONABLY_DISPLAYABLE_VERTICALLY:
+                    violations.append((_file_or_dir, len(_files)))
                 yield _file_or_dir
                 yield from walk(_file_or_dir)
             else:
                 yield _file_or_dir.resolve()
+        if violations:
```

Comment:
I said this before, why not just assert and do this at the end? This makes the code a lot more complex and harder to read.

Reply:
It makes the error friendlier for users and easier to tests. We have to do it at the end if we want to cover everything, or the first violations prevent from testing the other use case.